sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -315,18 +322,49 @@ def release(self, instruction_id):
     """
     descriptor_id, processor = self.active_bundle_processors.pop(instruction_id)
     processor.reset()
+    self.cached_bundle_processors_last_access_time[descriptor_id] = time.time()
     self.cached_bundle_processors[descriptor_id].append(processor)
 
   def shutdown(self):
     """
     Shutdown all ``BundleProcessor``s in the cache.
     """
+    if self.periodic_shutdown:
+      self.periodic_shutdown.cancel()
+      self.periodic_shutdown.join()
+      self.periodic_shutdown = None
+
     for instruction_id in self.active_bundle_processors:
       self.active_bundle_processors[instruction_id][1].shutdown()
       del self.active_bundle_processors[instruction_id]
     for cached_bundle_processors in self.cached_bundle_processors.values():
-      while len(cached_bundle_processors) > 0:
-        cached_bundle_processors.pop().shutdown()
+      BundleProcessorCache._shutdown_cached_bundle_processors(
+          cached_bundle_processors)
+
+  def _schedule_periodic_shutdown(self):
+    def shutdown_inactive_bundle_processors():
+      for descriptor_id, last_access_time in \
+              self.cached_bundle_processors_last_access_time.items():
+        if time.time() - last_access_time > \
+                DEFAULT_BUNDLE_PROCESSOR_CACHE_THRESHOLD_S:
```

Comment:
nit: you could surround the condition with paranthesis which would make it easier to read, e.g. 

```python
       if (time.time() - last_access_time >
                DEFAULT_BUNDLE_PROCESSOR_CACHE_THRESHOLD_S):
```